Topics/Tech_Stacks/Python_Requests.md

```diff
@@ -0,0 +1,128 @@
+}
+```
+You can then retrieve the content of the webpage, which in most cases is the HTML of the webpage.
+```python
+content = response.text
+```
+If you are retrieving JSON content from an API, you can use the corresponding `response.json()` method to output a dictionary.
+```python
+json_data = response.json()
+
+for key in json_data:
+  print(f"{key}: {json_data[key]}")
+```
+
+### POST
+Some endpoints may use a HTML Form or a JSON object to set the necessary parameters to display the data that you need. In ths case, you can use the `POST` method.
+```python
+data = {
+  'key': 'value'
+}
+
+# Send data as HTML form
+response = requests.post(url, data=data)
+
+# Send data as JSON
+response = requests.post(url, json=data)
+```
+You can then utilize `response` in the same way as `GET`.
+
+To determine the data that you need to send, you can use the Network tab in your browser's developer tools to examine the specific `POST` request that website sends.
+
+### Sessions
+Alternatively, you can use `SESSIONS` when making multiple requests to the same webpage, or if you want to persist cookies or other information like headers between requests. The syntax is largely the same with the exception of creating and using a session for making requests.
+```python
+session = requests.Session()
+response1 = session.post(url)
+response2 = session.get(url)
+```
+
+### User-Agent
+Some websites might employ protection against web scrapers. In those cases, you can add a User-Agent to send the request like a browser.
+```python
+headers = {
+    'User-Agent': 'My User Agent 1.0',
+}
+
+response = requests.get(url, headers=headers)
+
+# When using Sessions
+session = requests.Session()
+session.headers.update({'User-Agent': 'My User Agent 1.0'})
+```
+Some examples of user agents can be found [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent).
+
+### Status Code
+After executing the request, you can check the `status_code` of the request.
+```python
+if response.status_code == 200:
+  # success
+```
+Responses are classified into five groups:
+- `100` level - Informational: the request was received and the process is being continued
+- `200` level - Successful: the request was successful
+- `300` level - Redirection: the request requires further action
+- `400` level - Client Error: the request contains errors or could not be fulfilled
+- `500` level - Server Error: the server could not fulfill the request
+
+In most cases, you only have to worry about whether the status code matched a value of `200` or not.
+
+Alternatively, you can use Request's built-in `raise_for_status()`, which will raise an `HTTPError` if an unsuccessful status code was returned.
+```python
+try:
+  response.raise_for_status()
+except requests.exceptions.HTTPError as e:
+  # handle error here
+```
+
+A full list of HTTP response codes can be found [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status).
+
+
+## Parsing
```

Comment:
What is the point of parsing? What does the page look like after we parsed the data? Can you briefly explain beautifulsoup and attach a link to our page that explains the tool?

Reply:
Thanks, I will expand the section and include a link to a more detailed tutorial